# Module 02 Homework - Pandas
Using a dataset from _Wine Spectator_, a wine magazine, we will practice the material that we covered in notebook M02_Colab04 to M02_Colab06. This notebook will cover data transformation, grouping, and sorting using pandas.

Created by Lexinejazly Asuncion (ID #017077242)
Last updated: Sunday February 22, 2026

In [1]:
csvurl="https://gist.githubusercontent.com/clairehq/" + \
        "79acab35be50eaf1c383948ed3fd1129/raw/407a02139ae1e134992b90b4b2b8c329b3d73a6a/winemag-data-130k-v2.csv"
import pandas as pd
wine = pd.read_csv(csvurl)

In [7]:
wine.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

In [2]:
wine.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65499 entries, 0 to 65498
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             65499 non-null  int64  
 1   country                65467 non-null  object 
 2   description            65499 non-null  object 
 3   designation            46588 non-null  object 
 4   points                 65499 non-null  int64  
 5   price                  60829 non-null  float64
 6   province               65467 non-null  object 
 7   region_1               54744 non-null  object 
 8   region_2               25170 non-null  object 
 9   taster_name            51856 non-null  object 
 10  taster_twitter_handle  49467 non-null  object 
 11  title                  65499 non-null  object 
 12  variety                65499 non-null  object 
 13  winery                 65499 non-null  object 
dtypes: float64(1), int64(2), object(11)
memory usage: 7.0+

**Data cleaning**  
Notice that the first column is redundant. Part of data analysis is cleaning and removing redundancy. How would you drop the redundant column inplace, that is overwrite the dataframe.

#### Question 1: ####  
What is the mean of the points column?

In [43]:
#Find the statistics for each column
wine['points'].mean()

np.float64(88.43403716087269)

**The mean of the points column is about 88.43.**

#### Question 2: ####  
How many countries are present in this dataset? (Only count each country once)

In [12]:
#Select the country column and find the distinct number of values
wine['country'].nunique()

41

**There are 41 countries present in the dataset.**

#### Question 3: ####
How many times does each country appeared in this dataset? Show each country and the corresponding count (show counts in ascending order)

In [13]:
#Select the country column and count the number of rows with a unique value
wine['country'].value_counts(ascending=True)

country
Bosnia and Herzegovina        1
Slovakia                      1
Armenia                       1
Luxembourg                    4
India                         4
Switzerland                   4
Ukraine                       5
Czech Republic                6
Macedonia                     6
Cyprus                        6
Serbia                        7
Peru                          8
Morocco                      11
Lebanon                      20
Moldova                      30
Mexico                       31
Brazil                       31
England                      36
Georgia                      37
Slovenia                     39
Turkey                       43
Croatia                      44
Hungary                      61
Uruguay                      61
Romania                      67
Bulgaria                     68
Canada                      108
Greece                      242
Israel                      259
New Zealand                 733
South Africa                737
Germany                    1051
Australia                  1177
Austria                    1635
Argentina                  1907
Chile                      2258
Portugal                   2963
Spain                      3409
Italy                     10005
France                    11174
US                        27177
Name: count, dtype: int64

#### Question 4: ####
Create a variable `adjusted_price` containing the adjusted price which is the price subtracted by the average price. *This is called **"centering" transformation** - a method commonly used in the preprocessing step before applying various machine learning algorithms.*

In [35]:
#Compute the adjusted price for all the wine entries in the dataset
adjusted_price = wine['price'] - wine['price'].mean()
adjusted_price

0              NaN
1       -20.232932
2       -21.232932
3       -22.232932
4        29.767068
           ...    
65494     9.767068
65495   -13.232932
65496   -15.232932
65497    -4.232932
65498   -25.232932
Name: price, Length: 65499, dtype: float64

#### Question 5: ####
What is the title of the wine that has the highest points-to-price ratio in the dataset?

In [36]:
#Compute the points to price ratio for all wine entries in the dataset
wine['points_to_price'] = wine['points'] / wine['price']

#Locate and return the row index of the wine with the highest ratio, then select the wine title
wine.iloc[wine['points_to_price'].idxmax()]['title']


'Bandit NV Merlot (California)'

**The wine with the highest points to price ratio is Bandit NV Merlot (California).**

#### Question 6: ####
Create a series `flavor_counts` that contains two values: the number of wines that has the word "tart" in the `description` column and the number of wines that has the word "berries" in the `description` column. The index of the Series should be "Tart" and "Berries" for the corresponding values.

In [49]:
# Look for the word "tart", not case sensitive, not a substring of another word
# Same applies to the word "berries"

#Find the total number of the words "tart" and "berries" appear in the description column
tart_count = wine['description'].str.contains("tart", case=False, regex=False).sum()
berries_count = wine['description'].str.contains("berries", case=False, regex=False).sum()

#Make a series that has the counts with indices
flavor_counts = pd.Series([tart_count, berries_count], index=["Tart", "Berries"])
flavor_counts

Tart       4424
Berries    3512
dtype: int64

#### Question 7: ####
Let's convert the points into simple star ratings. A score of 90 or higher counts as 3 stars, a score of at least 80 but less than 90 is 2 stars. Any other score is 1 star.

Also, any wines from France should automatically get 3 stars, regardless of points.

Add this new column `star_ratings` to the dataframe with the number of stars for each wine in the dataset.

In [58]:
#Function to rate wines
def rate_wine(x):
  if x['country'] == "France":
    return 3
  elif x['points'] >= 90:
    return 3
  elif x['points'] >= 80:
    return 2
  else:
    return 1

#Apply the function to the DF, use axis=1 so the operation is performed row wise
wine["star_ratings"] = wine.apply(rate_wine, axis=1)
wine.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  points_to_price  star_ratings  
0     White Blend              Nicosia              NaN             2  
1  Portuguese Red  Quinta dos Avidagos         5.800000             2  
2      Pinot Gris            Rainstorm         6.214286             2  
3        Riesling           St. Julian         6.692308             2  
4      Pinot Noir         Sweet Cheeks         1.338462             2

#### Question 8: ####
Who are the most common wine reviewers in the dataset? Create a Series whose index is the taster_twitter_handle category from the dataset, and whose values count how many reviews each person wrote.

In [59]:
#Select the taster twitter handle column and count the number of rows for each reviewer
wine['taster_twitter_handle'].value_counts()

taster_twitter_handle
@vossroger          13045
@wineschach          7752
@kerinokeefe         5313
@paulgwine           4851
@vboone              4696
@mattkettmann        3035
@JoeCz               2605
@wawinereport        2358
@gordone_cellars     2032
@AnneInVino          1769
@laurbuzz             938
@suskostrzewa         593
@worldwineguys        465
@bkfiona               11
@winewchristina         4
Name: count, dtype: int64

**The most common wine reviewers in the dataset are @vossroger, @wineschach, and @kerinokeefe.**

#### Question 9: ####
What combination of countries and varieties are most common? Create a Series whose index is a MultiIndexof {country, variety} pairs. For example, a pinot noir produced in the US should map to {"US", "Pinot Noir"}. Sort the values in the Series in descending order based on wine count.

In [64]:
#Select the country and variety columns and count the number of unique combinations
combinations = wine.value_counts(['country', 'variety'])
combinations

country    variety                 
US         Pinot Noir                  4918
           Cabernet Sauvignon          3649
           Chardonnay                  3412
France     Bordeaux-style Red Blend    2380
Italy      Red Blend                   1870
                                       ... 
Romania    Rosé                           1
US         Ugni Blanc                     1
           Touriga                        1
           Torrontés                      1
Argentina  Merlot-Cabernet Franc          1
Name: count, Length: 1304, dtype: int64

**The most common combination of countries and varieties is US Pinot Noir.**

#### Question 10 #####
Create a Series whose index is reviewers and whose values is the average score given out by that reviewer.  
*Hint:* You will need the `taster_name` and `points` columns.

In [71]:
#Group the series by the reviewers name, then find the mean of the points given out for each
wine.groupby('taster_name').points.mean()

taster_name
Alexander Peartree    86.014286
Anna Lee C. Iijima    88.380506
Anne Krebiehl MW      90.587903
Carrie Dykes          86.644444
Christina Pickard     89.500000
Fiona Adams           87.090909
Jeff Jenssen          88.273504
Jim Gordon            88.604331
Joe Czerwinski        88.519770
Kerin O’Keefe         88.827969
Lauren Buzzeo         87.831557
Matt Kettmann         90.021087
Michael Schachner     86.904541
Mike DeSimone         89.030303
Paul Gregutt          89.095032
Roger Voss            88.678957
Sean P. Sullivan      88.666243
Susan Kostrzewa       86.408094
Virginie Boone        89.229557
Name: points, dtype: float64